### PR TITLE
[kernel] fix sliding window in prefix prefill Triton kernel

### LINKED ISSUE #4057: [Feature][Chunked prefill]: Make sliding window work
### 🚀 The feature, motivation and pitch

Currently, both prefix caching and chunked prefill doesn't work with sliding window attention because the block tables are not updated properly. Both uses `context_attention_forward` (https://github.com/vllm-project/vllm/blob/5c2e66e4871917c5d59cc4a8b89ef53e690e9bd9/vllm/attention/ops/prefix_prefill.py#L621), which I am not sure if sliding window is supported. We should fix this if we want to enable chunked prefill/prefix caching for models that use sliding window attn (e.g., mistral 7B). 

### Alternatives

N/A

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mmoskal**
This adds support for the sliding window in prefix prefill kernel.

I had to use a large negative value instead of -inf for masking, since otherwise in some situations we get '-inf - -inf' in softmax which leads to NaNs.

Added tests comparing with xformers.

Also added a bunch of comments with tensor shapes etc.

FIX #4057

CC @rkooo567 @cadedaniel @simon-mo 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by zhuohan123** [comment]
CC @caoshiyi Can you help take a look at this?

**Comment by rkooo567** [comment]
yep! it is Sat here, but I will take a look at it very soon! @mmoskal thanks for the amazing contribution!

File: vllm/attention/ops/prefix_prefill.py [hunk 1]
```diff
@@ -62,42 +63,53 @@ def _fwd_kernel(
         cur_batch_in_all_start_index = tl.load(B_Start_Loc + cur_batch)
         cur_batch_query_len = cur_batch_seq_len - cur_batch_ctx_len
 
+        # start position inside of the query
+        # generally, N goes over ctx_len or seq_len, while M goes over query_len
```
Comment (rkooo567) [inline vllm/attention/ops/prefix_prefill.py:67]:
```suggestion
        # generally, N goes over kv, while M goes over query_len
```
? I think the definition of ctx_len is not very clear in vllm

Comment (rkooo567) [reply vllm/attention/ops/prefix_prefill.py:67]: for our input, ctx_len means the key size and seq_len means the key + query size

File: vllm/attention/ops/prefix_prefill.py [hunk 2]
```diff
@@ -106,23 +118,32 @@ def _fwd_kernel(
...
             qk += tl.dot(q, k)
             qk = tl.where((start_n + offs_n[None, :]) < cur_batch_ctx_len, qk,
                           float("-inf"))
+            if SLIDING_WINDOW > 0:
+                # We can't use -inf here, because the
+                # sliding window may lead to the entire row being masked.
+                # This then makes m_ij contain -inf, which causes NaNs in
+                # exp().
+                qk = tl.where((cur_batch_ctx_len + offs_m[:, None]) -
+                              (start_n + offs_n[None, :]) < SLIDING_WINDOW, qk,
```
Comment (rkooo567) [inline vllm/attention/ops/prefix_prefill.py:133]: so this means 
```
cur_batch_ctx_len + offs_m[:, None] == end of q
start_n + offs_n[None, :] == end of k
```

so q-k length is within slinding window, attend it, is this correct?

**Review by rkooo567 — COMMENTED** [review]
Looks pretty good! Also thanks for the comment on shapes. Some comments regarding tests

File: tests/kernels/test_prefix_prefill.py [hunk 3]
```diff
@@ -15,18 +15,21 @@
 CUDA_DEVICES = [
     f"cuda:{i}" for i in range(1 if torch.cuda.device_count() == 1 else 2)
 ]
+SLIDING_WINDOW = [0, 512]
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply tests/kernels/test_prefix_prefill.py:18]: Also curious if we can do e2e test against mistral or other model that has sliding window attn enabled..

[the remark this answers is not in the document]
Comment (mmoskal) [reply tests/kernels/test_prefix_prefill.py:18]: I'll try different sizes. I'm trying to get the sliding window to work with v2 block manager (somewhat based on #3967) which should exercise this.

File: vllm/attention/ops/prefix_prefill.py [hunk 4]
```diff
@@ -106,23 +118,32 @@ def _fwd_kernel(
...
+                              -10000)
```
[the remark this answers is not in the document]
Comment (mmoskal) [reply vllm/attention/ops/prefix_prefill.py:134]: -65000 or so is the smallest value for f16. Anyway, this gets exp()ed so -10k should enough.

File: tests/kernels/test_prefix_prefill.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rkooo567) [reply tests/kernels/test_prefix_prefill.py:18]: sg. The test itself will be done in this PR right?

File: vllm/attention/ops/prefix_prefill.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mmoskal) [reply vllm/attention/ops/prefix_prefill.py:134]: Actually good you point this out - I found that in one place I do sm_scale before and in the other after masking with -10000; it doesn't matter much (smallest scale is 1/16 or so, and exp(-10000/16) is still zero) but better to always scale and then mask so we get consistent mask values - fixed

File: tests/kernels/test_prefix_prefill.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmoskal) [reply tests/kernels/test_prefix_prefill.py:18]: yes, adding more sliding window parameters now

**Review by rkooo567 — APPROVED** [review]
LGTM given it passes with more variations in unit test!

File: vllm/attention/ops/prefix_prefill.py (hunk 2 above)
Comment (mmoskal) [reply vllm/attention/ops/prefix_prefill.py:139]: yes; I added a comment to that effect

**Comment by mmoskal** [comment]
@rkooo567 all tests passed, should be good

**Comment by rkooo567** [comment]
cc @simon-mo to merge!

**Comment by mmoskal** [comment]
@simon-mo can we get this merged? I need it to rebase #4545

**Review by simon-mo — APPROVED** [review]

**Pull request merged**